src/test/util/setup_common.cpp

```diff
@@ -109,7 +109,18 @@ TestingSetup::TestingSetup(const std::string& chainName) : BasicTestingSetup(cha
     GetMainSignals().RegisterBackgroundSignalScheduler(scheduler);
 
     pblocktree.reset(new CBlockTreeDB(1 << 20, true));
-    g_chainstate = MakeUnique<CChainState>();
+
+    // During QT tests, we may end up calling this twice (e.g. during RPCNestedTests)
+    // which throws an excetion. Chainstates should otherwise never be overwritten
+    // during normal operation.
+    try {
+        g_chainman.InitializeChainstate();
+    } catch (const std::logic_error& e) {
+        std::string msg = e.what();
+        if (msg.find("overwriting a chainstate") == std::string::npos) {
+            throw;
+        }
+    }
```

Comment:
Please remove this hack in the unit tests. It will lead to potentially not resetting global state and preserving a chainstate of previous tests. The test failure is due to a missing reset in the gui tests, which can be fixed with a one-line patch:


```diff
diff --git a/src/qt/test/apptests.cpp b/src/qt/test/apptests.cpp
index 14a75b23f3..f9eb4cde30 100644
--- a/src/qt/test/apptests.cpp
+++ b/src/qt/test/apptests.cpp
@@ -82,6 +82,7 @@ void AppTests::appTests()
     // Reset global state to avoid interfering with later tests.
     AbortShutdown();
     UnloadBlockIndex();
+    g_chainman.Reset();
 }
 
 //! Entry point for BitcoinGUI tests.

Reply:
Ah, thanks for the patch. Fixed.